File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_READMEMORY.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_READMEMORY : Message
    {
        public SID_READMEMORY()
        {
            Id = (byte)MessageIds.SID_READMEMORY;
            Buffer = new byte[0];
        }

        public SID_READMEMORY(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_READMEMORY;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

                        /**
                         * (UINT32) Request ID
                         * (VOID) Memory
                         */

                        if (Buffer.Length < 4)
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 4 bytes");

                        using var m = new MemoryStream(Buffer);
                        using var r = new BinaryReader(m);

                        var requestId = r.ReadUInt32();
                        var data = r.ReadBytes((int)(r.BaseStream.Length - r.BaseStream.Position));

                        // We don't use this. Why did the client send us this ??

                        break;
                    }
                case MessageDirection.ServerToClient:
                    {
                        var requestId = (UInt32)context.Arguments["requestId"];
                        var address = (UInt32)context.Arguments["address"];
                        var length = (UInt32)context.Arguments["length"];

                        /**
                         * (UINT32) Request ID
                         * (UINT32) Address
                         * (UINT32) Length
                         */

                        Buffer = new byte[12];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write((UInt32)requestId);
                        w.Write((UInt32)address);
                        w.Write((UInt32)length);

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_READMEMORY.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_READMEMORY
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_READMEMORY)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_READMEMORY)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    If Buffer.Length < 4 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 4 bytes")

                    Using m = New MemoryStream(Buffer)
                        Using r = New BinaryReader(m)
                            Dim requestId = r.ReadUInt32()
                            Dim data = r.ReadBytes(CInt((r.BaseStream.Length - r.BaseStream.Position)))
                        End Using
                    End Using

                    Exit Select
                Case MessageDirection.ServerToClient
                    Dim requestId = CUInt(context.Arguments("requestId"))
                    Dim address = CUInt(context.Arguments("address"))
                    Dim length = CUInt(context.Arguments("length"))
                    Buffer = New Byte(11) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(CUInt(requestId))
                            w.Write(CUInt(address))
                            w.Write(CUInt(length))
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
